src/anomalib/models/ai_vad/features.py

```diff
@@ -0,0 +1,239 @@
+
+    POSE = "pose"
+    VELOCITY = "velocity"
+    DEEP = "deep"
+
+
+class FeatureExtractor(nn.Module):
+    """Feature extractor for AI-VAD.
+
+    Args:
+        n_velocity_bins (int): Number of discrete bins used for velocity histogram features.
+        use_velocity_features (bool): Flag indicating if velocity features should be used.
+        use_pose_features (bool): Flag indicating if pose features should be used.
+        use_deep_features (bool): Flag indicating if deep features should be used.
+    """
+
+    def __init__(
+        self,
+        n_velocity_bins: int = 8,
+        use_velocity_features: bool = True,
+        use_pose_features: bool = True,
+        use_deep_features: bool = True,
+    ) -> None:
+        super().__init__()
+        assert (
+            use_velocity_features or use_pose_features or use_deep_features
+        ), "At least one feature stream must be enabled."
+
+        self.use_velocity_features = use_velocity_features
+        self.use_pose_features = use_pose_features
+        self.use_deep_features = use_deep_features
+
+        self.appearance_extractor = DeepExtractor()
+        self.velocity_extractor = VelocityExtractor(n_bins=n_velocity_bins)
+        self.pose_extractor = PoseExtractor()
+
+    def forward(
+        self,
+        rgb_batch: Tensor,
+        flow_batch: Tensor,
+        regions: list[dict],
+    ) -> list[dict]:
+        """Forward pass through the feature extractor.
+
+        Extract any combination of velocity, pose and deep features depending on configuration.
+
+        Args:
+            rgb_batch (Tensor): Batch of RGB images of shape (N, 3, H, W)
+            flow_batch (Tensor): Batch of optical flow images of shape (N, 2, H, W)
+            regions (list[dict]): Region information per image in batch.
+        Returns:
+            list[dict]: Feature dictionary per image in batch.
+        """
+        batch_size = rgb_batch.shape[0]
+
+        # convert from list of [N, 4] tensors to single [N, 5] tensor where each row is [index-in-batch, x1, y1, x2, y2]
+        boxes_list = [batch_item["boxes"] for batch_item in regions]
+        indices = torch.repeat_interleave(
+            torch.arange(len(regions)), Tensor([boxes.shape[0] for boxes in boxes_list]).int()
+        )
+        boxes = torch.cat([indices.unsqueeze(1).to(rgb_batch.device), torch.cat(boxes_list)], dim=1)
+
+        # Extract features
+        feature_dict = {}
+        if self.use_velocity_features:
+            velocity_features = self.velocity_extractor(flow_batch, boxes)
+            feature_dict[FeatureType.VELOCITY] = [velocity_features[indices == i] for i in range(batch_size)]
+        if self.use_pose_features:
+            pose_features = self.pose_extractor(rgb_batch, boxes_list)
+            feature_dict[FeatureType.POSE] = pose_features
+        if self.use_deep_features:
+            deep_features = self.appearance_extractor(rgb_batch, boxes, batch_size)
```

Comment:
Is `deep_features` the way how it is named in the paper?  Or `appearance_features`?

Reply:
I'm a bit confused here. `self.appearance_extractor` is a `DeepExtractor` object. I feel like the naming would be more consistent?

Reply:
I changed it from `appearance` to `deep`, to make it consistent with the paper, but missed this attribute apparently